src/comments.js

```diff
@@ -252,42 +283,90 @@ function addTrailingComment(node, comment) {
+function printComment(commentPath) {
+  const comment = commentPath.getValue();
+
+  switch(comment.type) {
+    case "CommentBlock":
+    case "Block":
+      return "/*" + comment.value + "*/";
+    case "CommentLine":
+    case "Line":
+      return "//" + comment.value;
+    default:
+      throw new Error("Not a comment: " + JSON.stringify(comment));
+  }
 }
 
-function printTrailingComment(commentPath, print, options) {
-  const comment = commentPath.getValue(commentPath);
-  n.Comment.assert(comment);
+function printLeadingComment(commentPath, print, options) {
+  const comment = commentPath.getValue();
+  const contents = printComment(commentPath);
   const text = options.originalText;
+  const isBlock = comment.type === "Block" || comment.type === "CommentBlock";
+
+  // Leading block comments should see if they need to stay on the
+  // same line or not.
+  if(isBlock) {
+    return concat([
+      contents,
+      util.hasNewline(options.originalText, locEnd(comment)) ? hardline : " "
+    ]);
+  }
+
+  return concat([contents, hardline]);
+}
+
+function printTrailingComment(commentPath, print, options, parentNode) {
+  const comment = commentPath.getValue();
+  const contents = printComment(commentPath);
+  const isBlock = comment.type === "Block" || comment.type === "CommentBlock";
+
+  if(util.hasNewline(options.originalText, locStart(comment), true)) {
+    // This allows comments at the end of nested structures:
+    // {
+    //   x: 1,
+    //   y: 2
+    //   // A comment
+    // }
+    // Those kinds of comments are almost always leading comments, but
+    // here it doesn't go "outside" the block and turns it into a
+    // trailing comment for `2`. We can simulate the above by checking
+    // if this a comment on its own line; normal trailing comments are
+    // always at the end of another expression.
+    return concat([
+      hardline,
+      contents
+    ]);
+  } else if (isBlock) {
+    // Trailing block comments never need a newline
+    return concat([" ", contents]);
+  }
 
   return concat([
-    util.newlineExistsBefore(text, locStart(comment)) ? hardline : " ",
-    print(commentPath)
-  ]);
+    lineSuffix(" " + contents),
+    !isBlock ? breakParent : ""
+  ])
 }
 
 function printDanglingComments(path, print, options) {
   const text = options.originalText;
-
   const parts = [];
-  path.each(
-    commentPath => {
-      const comment = commentPath.getValue();
-      if (!comment.leading && !comment.trailing) {
-        parts.push(
-          util.newlineExistsBefore(text, locStart(comment)) ? hardline : " "
-        );
-        parts.push(commentPath.call(print));
-      }
-    },
-    "comments"
-  );
+  path.each(commentPath => {
+    const comment = commentPath.getValue();
+    if(!comment.leading && !comment.trailing) {
+      parts.push(hardline);
+      parts.push(printComment(commentPath));
+    }
+  }, "comments");
   return concat(parts);
 }
 
+function isList(node) {
```

Comment:
Yep! This was from a previous exploration, which I abandoned for exactly that reason. Will remove.